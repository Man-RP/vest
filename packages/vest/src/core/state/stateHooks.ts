import ctx from 'ctx';

// @ts-ignore
const getStateRef = () => ctx.use().stateRef;

export function usePending() {
  // @ts-ignore
  return getStateRef().pending();
}
export function useSuiteId() {
  // @ts-ignore
  return getStateRef().suiteId();
}
export function useTestCallbacks() {
  // @ts-ignore
  return getStateRef().testCallbacks();
}
export function useTestObjects() {
  // @ts-ignore
  return getStateRef().testObjects();
}
export function useSkippedTests() {
  // @ts-ignore
  return getStateRef().skippedTests();
}
export function useOptionalFields() {
  // @ts-ignore
  return getStateRef().optionalFields();
}
