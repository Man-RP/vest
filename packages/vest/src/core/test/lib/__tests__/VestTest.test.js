import runCreateRef from '../../../../../testUtils/runCreateRef';

import VestTest from 'VestTest';
import addTestToState from 'addTestToState';
import context from 'ctx';
import { setPending } from 'pending';
import { usePending, useTestObjects } from 'stateHooks';
import * as testStatuses from 'testStatuses';

const fieldName = 'unicycle';
const statement = 'I am Root.';

let stateRef;

it.ctx = (str, cb) => it(str, () => context.run({ stateRef }, cb));

describe('VestTest', () => {
  let testObject;

  beforeEach(() => {
    testObject = VestTest({
      fieldName,
      statement,
      testFn: jest.fn(),
    });
  });

  test('TestObject constructor', () => {
    expect(testObject).toMatchSnapshot();
  });

  it('Should have a unique id', () => {
    Array.from({ length: 100 }, () =>
      VestTest({ fieldName, statement, testFn: jest.fn() })
    ).reduce((existing, { id }) => {
      expect(existing[id]).toBeUndefined();
      existing[id] = true;
      return existing;
    }, {});
  });

  describe('testObject.warn', () => {
    it('Should set `.isWarning` to true', () => {
      expect(testObject.isWarning).toBe(false);
      testObject.warn();
      expect(testObject.isWarning).toBe(true);
      expect(testObject).toMatchSnapshot();
    });
  });

  describe('testObject.fail', () => {
    beforeEach(() => {
      jest.resetModules();

      const VestTest = require('VestTest');
      testObject = VestTest({
        fieldName,
        statement,
        testFn: jest.fn(),
      });
    });

    afterEach(() => {
      jest.resetAllMocks();
    });

    it('Should set this.status to FAILED', () => {
      expect(testObject.status).not.toBe(testStatuses.FAILED);
      testObject.fail();
      expect(testObject.status).toBe(testStatuses.FAILED);
    });
  });

  describe('testobject.valueOf', () => {
    test('When `failed` is false', () => {
      expect(testObject.status).not.toBe(testStatuses.FAILED);
      expect(testObject.valueOf()).toBe(true);
    });

    test('When `failed` is true', () => {
      testObject.status = testStatuses.FAILED;
      expect(testObject.valueOf()).toBe(false);
    });
  });

  describe('testObject.cancel', () => {
    beforeEach(() => {
      stateRef = runCreateRef();

      context.run({ stateRef }, () => {
        addTestToState(testObject);
      });
    });
    it.ctx('Should remove a testObject from the state', () => {
      const [testObjects] = useTestObjects();
      expect(testObjects).toEqual(expect.arrayContaining([testObject]));
      testObject.cancel();
      expect(testObjects).toEqual(expect.not.arrayContaining([testObject]));
    });

    it.ctx('Should remove a testObject from the pending state', () => {
      setPending(testObject);
      {
        const [{ pending }] = usePending();
        expect(pending).toEqual(expect.arrayContaining([testObject]));
      }
      testObject.cancel();
      {
        const [{ pending }] = usePending();
        expect(pending).toEqual(expect.not.arrayContaining([testObject]));
      }
    });

    it.ctx('Should remove a testObject from the lagging state', () => {
      const [, setPending] = usePending();
      setPending(state => ({ ...state, lagging: [testObject] }));
      {
        const [{ lagging }] = usePending();
        expect(lagging).toEqual(expect.arrayContaining([testObject]));
      }
      testObject.cancel();
      {
        const [{ lagging }] = usePending();
        expect(lagging).toEqual(expect.not.arrayContaining([testObject]));
      }
    });
  });
});
