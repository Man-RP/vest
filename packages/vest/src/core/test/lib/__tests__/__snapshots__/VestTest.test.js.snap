// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VestTest TestObject constructor 1`] = `
Object {
  "cancel": [Function],
  "fail": [Function],
  "fieldName": "unicycle",
  "id": "0",
  "isWarning": false,
  "setStatus": [Function],
  "statement": "I am Root.",
  "status": "untested",
  "testFn": [MockFunction],
  "valueOf": [Function],
  "warn": [Function],
}
`;

exports[`VestTest testObject.warn Should set \`.isWarning\` to true 1`] = `
Object {
  "cancel": [Function],
  "fail": [Function],
  "fieldName": "unicycle",
  "id": "102",
  "isWarning": true,
  "setStatus": [Function],
  "statement": "I am Root.",
  "status": "untested",
  "testFn": [MockFunction],
  "valueOf": [Function],
  "warn": [Function],
}
`;
